src/main/java/org/folio/service/claiming/ClaimingService.java

```diff
@@ -0,0 +1,246 @@
+            }
+            return createJobsByVendor(config, pieceIdsByVendorIds, requestContext);
+          });
+      })
+      .onFailure(t -> logger.error("sendClaims:: Failed send claims: {}", JsonObject.mapFrom(claimingCollection).encodePrettily(), t));
+  }
+
+  private Future<Map<String, List<String>>> groupPieceIdsByVendorId(List<String> pieceIds, RequestContext requestContext) {
+    if (CollectionUtils.isEmpty(pieceIds)) {
+      logger.info("groupPieceIdsByVendorId:: No pieces are grouped by vendor, pieceIds is empty");
+      return Future.succeededFuture();
+    }
+    logger.info("groupPieceIdsByVendorId:: Grouping pieces by vendor, pieceIds count: {}", pieceIds.size());
+    return pieceStorageService.getPiecesByIds(pieceIds, requestContext)
+      .compose(pieces -> {
+        if (CollectionUtils.isEmpty(pieces)) {
+          logger.info("groupPieceIdsByVendorId:: No pieces are found by piece ids, pieceIds: {}", pieceIds);
+          return Future.succeededFuture();
+        }
+        var uniquePiecePoLinePairs = pieces.stream()
+          .filter(Objects::nonNull).filter(piece -> Objects.nonNull(piece.getId()) && Objects.nonNull(piece.getPoLineId()))
+          .map(piece -> Pair.of(piece.getPoLineId(), piece.getId())).distinct()
+          .toList();
+        logger.info("groupPieceIdsByVendorId:: Prepared unique piece-poLine pairs, pairs: {}", uniquePiecePoLinePairs);
+        return collectResultsOnSuccess(createPieceIdByVendorFutures(pieces, uniquePiecePoLinePairs, requestContext))
+          .map(ClaimingService::transformAndGroupPieceIdsByVendorId);
+      });
+  }
+
+  private List<Future<Pair<String, String>>> createPieceIdByVendorFutures(List<Piece> pieces, List<Pair<String, String>> uniquePiecePoLinePairs,
+                                                                          RequestContext requestContext) {
+    var pieceIdByVendorIdFutures = new ArrayList<Future<Pair<String, String>>>();
+    uniquePiecePoLinePairs.forEach(piecePoLinePairs -> {
+      var foundPiece = pieces.stream()
+        .filter(Objects::nonNull).filter(piece -> Objects.nonNull(piece.getId())).filter(piece -> piece.getId().equals(piecePoLinePairs.getRight()))
+        .findFirst().orElseThrow(() -> new NoSuchElementException(String.format(CANNOT_FIND_A_PIECE_BY_ID, piecePoLinePairs.getRight())));
+      var pieceIdByVendorIdFuture = createVendorPiecePair(piecePoLinePairs, foundPiece, requestContext);
+      if (Objects.nonNull(pieceIdByVendorIdFuture)) {
+        pieceIdByVendorIdFutures.add(pieceIdByVendorIdFuture);
+      }
+    });
+    return pieceIdByVendorIdFutures;
+  }
+
+  private Future<Pair<String, String>> createVendorPiecePair(Pair<String, String> piecePoLinePairs,
+                                                             Piece piece, RequestContext requestContext) {
+    if (Objects.nonNull(piece) && !piece.getReceivingStatus().equals(Piece.ReceivingStatus.LATE)) {
```

Comment:
Currently null pieces won't be present here, so let's remove nonNull check or use OR operator - if piece is null this condition will be skipped